s3/pkg/service/multipart.go

```diff
@@ -0,0 +1,451 @@
+	contentType := multipart.Metadata.ContentType
+	object := &pb.Object{
+		BucketName:       bucketName,
+		ObjectKey:        objectKey,
+		TenantId:         multipart.Metadata.TenantId,
+		UserId:           multipart.Metadata.UserId,
+		ContentType:      contentType,
+		ObjectId:         multipart.ObjectId,
+		LastModified:     time.Now().UTC().Unix(),
+		Etag:             result.ETag,
+		DeleteMarker:     false,
+		CustomAttributes: multipart.Metadata.Attrs,
+		Type:             ObjectTypeNormal,
+		Tier:             utils.Tier1,
+		//TODO: update size, but the size should be returned from backend
+	}
+
+	err = s.MetaStorage.PutObject(ctx, &Object{Object: object}, &multipart, nil, true)
+	if err != nil {
+		log.Errorln("failed to put object meta. err:", err)
+		// TODO:  delete object
+		return ErrDBError
+	}
+
+	err = s.MetaStorage.DeleteMultipart(ctx, multipart)
+	if err != nil {
+		log.Errorln("failed to delete multipart. err:", err)
+		return err
+	}
+
+	log.Infoln("CompleteMultipartUpload upload part successfully.")
+	return nil
+}
+
+func (s *s3Service) AbortMultipartUpload(ctx context.Context, in *pb.AbortMultipartRequest, out *pb.BaseResponse) error {
+	log.Info("AbortMultipartUpload is called in s3 service.")
+	bucketName := in.BucketName
+	objectKey := in.ObjectKey
+	uploadId := in.UploadId
+
+	var err error
+	abortMultipartResponse := pb.BaseResponse{}
+	defer func() {
+		abortMultipartResponse.ErrorCode = GetErrCode(err)
+	}()
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, bucketName, true)
+	if err != nil {
+		log.Errorln("failed to get bucket from meta storage. err:", err)
+		return err
+	}
+
+	if !isAdmin {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId {
+				log.Errorln("bucket owner is not equal to request owner.")
+				return ErrBucketAccessForbidden
+			}
+		}
+	}
+
+	multipart, err := s.MetaStorage.GetMultipart(bucketName, objectKey, uploadId)
+	if err != nil {
+		log.Errorln("failed to get multipart info. err:", err)
+		return err
+	}
+
+	backendName := bucket.DefaultLocation
+	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage. err:", err)
+		return err
+	}
+
+	err = sd.AbortMultipartUpload(ctx, &pb.MultipartUpload{Bucket: bucketName, Key: objectKey, UploadId: uploadId, ObjectId: "objectId"})
+	if err != nil {
+		log.Errorln("failed to abort multipart. err:", err)
+		return err
+	}
+
+	err = s.MetaStorage.DeleteMultipart(ctx, multipart)
+	if err != nil {
+		log.Errorln("failed to delete multipart. err:", err)
+		return err
```

Comment:
Is it necessary to return error, or just let it to be cleaned by gc, and do not return error?

Reply:
the same as above